### PR TITLE
gh-96534: socketmodule: support FreeBSD divert(4) socket

### PR DESCRIPTION
**Pull request by glebius**
All rationale in #96534 

* Issue: gh-96534

### CODE REVIEW

**Comment by glebius** [comment]
@tiran @nirs @zooba @vstinner

Since this trivial pull request is being ignored for 2 months, tagging people who were the most recent committers to this module. Please take a look!

**Comment by zooba** [comment]
I see no problem with it, as it's just constants. But should we have a test?

**Comment by glebius** [comment]
> I see no problem with it, as it's just constants. But should we have a test?

Do you have FreeBSD in your CI? Is it worth to have one just for this small case?

I can ensure you that this code will be exercised in our FreeBSD CI, as we use python for regression test of the divert module :)
[This change](https://cgit.freebsd.org/src/commit/tests/sys/common/divert.py?id=4627bc1e90fd17c8d4f44de6b932d456b67661fc) just waits to be enabled back in our CI as soon as python accepts this pull request and new python version hits FreeBSD ports.

**Review by vstinner — CHANGES_REQUESTED** [review]
I don't get the point of this PR. ``AF_DIVERT`` and ``PF_DIVERT`` don't exist my FreeBSD 13 VM (grep doesn't match any line):

```
vstinner@freebsd$ grep AF_DIVERT -R /usr/include/ 
vstinner@freebsd$ grep PF_DIVERT -R /usr/include/ 
```

But IPPROTO_DIVERT is defined:

```
vstinner@freebsd$ grep IPPROTO_DIVERT -R /usr/include/ 
/usr/include/netinet/in.h:#define       IPPROTO_DIVERT          258             /* divert pseudo-protocol */
```

And the manual page mentioned in the Python issue contains:

```
int socket(PF_INET, SOCK_RAW, IPPROTO_DIVERT);
```

For me, IPPROTO_DIVERT constant should be added, no?

**Comment by vstinner** [comment]
> Do you have FreeBSD in your CI?

There are FreeBSD buildbot workers, but I don't think that it's worth it to add an unit test (I'm fine with merging the PR without a test).

**Comment by glebius** [comment]
> I don't get the point of this PR. `AF_DIVERT` and `PF_DIVERT` don't exist my FreeBSD 13 VM (grep doesn't match any line):
..
> For me, IPPROTO_DIVERT constant should be added, no?

The new protocol family exists only in FreeBSD 14 and the old PF_INET/IPPROTO_DIVERT will go away in FreeBSD 14:

https://cgit.freebsd.org/src/commit/sys/netinet/ip_divert.c?id=8624f4347e8133911b0554e816f6bedb56dc5fb3

All known open source software is already patched wrt this change.

The way I patched the Python's socketmodule allows it to be compiled both on FreeBSD 14 and old versions correctly, so that new builds will support PF_DIVERT and old builds will allow to use PF_INET with protocol integer equal to IPPROTO_DIVERT, just as they did before.

**Comment by glebius** [comment]
> FreeBSD divert manual page: https://www.freebsd.org/cgi/man.cgi?divert

[FreeBSD 14 divert manual page](https://www.freebsd.org/cgi/man.cgi?query=divert&apropos=0&sektion=0&manpath=FreeBSD+14.0-current&arch=default&format=html)

File: Modules/socketmodule.c
Comment (nirs) [inline Modules/socketmodule.c:7641]: It would be nice to add a comment why we need to check for both PF_DIVERT and
AF_DIVERT and mention the related freebsd versions.

You explained this in the discussion but having this in the code will make it much
easier to maintain 5 years from now.

File: Modules/socketmodule.c
Comment (nirs) [inline Modules/socketmodule.c:1856]: I wonder why we don't need to check for PF_DIVERT here? Do we need a comment for this?

**Comment by vstinner** [comment]
> FreeBSD 14 divert manual page

Ah, I tested your PR on a FreeBSD 13 VM. When is the FreeBSD 14 release planned?

I understand that his PR is not usable on FreeBSD 13, right? Would it be useful to expose IPPROTO_DIVERT for FreeBSD 13?

File: Misc/NEWS.d/next/Library/2022-09-03-09-24-02.gh-issue-96534.EU4Oxv.rst
Comment (nirs) [inline Misc/NEWS.d/next/Library/2022-09-03-09-24-02.gh-issue-96534.EU4Oxv.rst]: Only on 14?

**Comment by vstinner** [comment]
The current https://cgit.freebsd.org/src/plain/tests/sys/common/divert.py script mentioned in the issue uses IPPROTO_DIVERT (by defining it manually to 258).

**Comment by vstinner** [comment]
cc @koobs

**Comment by koobs** [comment]
> cc @koobs

We have my buildbot workers for 12.x and CURRENT, so QA/CI isn't an issue.

@glebius if this landed only recently (If its a more recent CURRENT than a few months), I'll need to update (there's been a zfs arc regression preventing upgrade) the worker first.

File: Modules/socketmodule.c
Comment (glebius) [reply Modules/socketmodule.c:7641]: The distinction between **protocol families** (PF_* constants) and **address families** (AF_* constants) is what we inherited from the original BSD sockets API. The PF_FOO is supposed to be an argument for socket(2), while AF_FOO for any socket address manipulation syscalls or functions, e.g. bind(), getsockname(), etc. This distinction makes some sense, since existence of a protocol doesn't dictate existence of address type and vice versa. The divert(4) is actually nice example, as it uses AF_INET addresses. Or look at AF_RDS a couple lines above in the socketmodule.c - the same.

**However**, all modern implementations of sockets always declare both AF_FOO and PF_FOO. As far as I know, they always are equal. POSIX [doesn't mention PF_* families](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_socket.h.html). Linux [notes their existence in the NOTES of the manual page](https://man7.org/linux/man-pages/man2/socket.2.html). Many software developers would use AF_FOO when calling socket(). This is why AF_DIVERT is also provided by FreeBSD and this is why I provided PyModule_AddIntMacro for it, too.

File: Modules/socketmodule.c
Comment (glebius) [reply Modules/socketmodule.c:1856]: Neither PF_DIVERT of AF_DIVERT are provided on old versions of FreeBSD. Both are provided on the new versions. I think this is stylistically more correct to put ifdefs as I did. For longer explanation on the difference between them, see my other comment.

**Comment by vstinner** [comment]
cc @jehops @emaste

**Comment by glebius** [comment]
> Ah, I tested your PR on a FreeBSD 13 VM. When is the FreeBSD 14 release planned?

It is planned for summer 2023.

> I understand that his PR is not usable on FreeBSD 13, right? Would it be useful to expose IPPROTO_DIVERT for FreeBSD 13?

Right, on FreeBSD 13 the new code won't be compiled in. I don't think there is any reason to add support for obsolete constant. To my knowledge our regression suite CI is the only Python code that uses it.

File: Modules/socketmodule.c
Comment (glebius) [reply Modules/socketmodule.c:7641]: The distinction between **protocol families** (PF_* constants) and **address families** (AF_* constants) is what we inherited from the original BSD sockets API. The PF_FOO is supposed to be an argument for socket(2), while AF_FOO for any socket address manipulation syscalls or functions, e.g. bind(), getsockname(), etc. This distinction makes some sense, since existence of a protocol doesn't dictate existence of address type and vice versa. The divert(4) is actually nice example, as it uses AF_INET addresses. Or look at AF_RDS a couple lines above in the socketmodule.c - the same.

**However**, all modern implementations of sockets always declare both AF_FOO and PF_FOO. As far as I know, they always are equal. POSIX [doesn't mention PF_* families](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_socket.h.html). Linux [notes their existence in the NOTES of the manual page](https://man7.org/linux/man-pages/man2/socket.2.html). Many software developers would use AF_FOO when calling socket(). This is why AF_DIVERT is also provided by FreeBSD and this is why I provided PyModule_AddIntMacro for it, too.

File: Modules/socketmodule.c
Comment (glebius) [reply Modules/socketmodule.c:1856]: Neither PF_DIVERT of AF_DIVERT are provided on old versions of FreeBSD. Both are provided on the new versions. I think this is stylistically more correct to put ifdefs as I did. For longer explanation on the difference between them, see my other comment.

File: Misc/NEWS.d/next/Library/2022-09-03-09-24-02.gh-issue-96534.EU4Oxv.rst
Comment (glebius) [reply Misc/NEWS.d/next/Library/2022-09-03-09-24-02.gh-issue-96534.EU4Oxv.rst]: Well, 14 and beyond to be precise.

File: Modules/socketmodule.c
Comment (nirs) [reply Modules/socketmodule.c:7641]: Right, looking at the rest of the code it is expected to have both constants.

File: Misc/NEWS.d/next/Library/2022-09-03-09-24-02.gh-issue-96534.EU4Oxv.rst
Comment (nirs) [reply Misc/NEWS.d/next/Library/2022-09-03-09-24-02.gh-issue-96534.EU4Oxv.rst]: Maybe "Support divert(4) added in FreeBSD 14."?

**Review by nirs — COMMENTED** [review]
Looks good to me.

It would be nice to have tests using the new constants when they are available.

File: Misc/NEWS.d/next/Library/2022-09-03-09-24-02.gh-issue-96534.EU4Oxv.rst
Comment (glebius) [reply Misc/NEWS.d/next/Library/2022-09-03-09-24-02.gh-issue-96534.EU4Oxv.rst]: Pushed that!

**Comment by koobs** [comment]
> For me, IPPROTO_DIVERT constant should be added, no?

Addressed in https://github.com/python/cpython/pull/96536#issuecomment-1301414704, no longer awaiting changes

**Comment by vstinner** [comment]
> I don't think there is any reason to add support for obsolete constant.

They are users running FreeBSD older than FreeBSD 14. Without IPPROTO_DIVERT exposed in Python, this kind of socket cannot be used in Python.

Well, if your plan is to only add support for divert sockets on FreeBSD 14, I will wait until FreeBSD 14 is released to manually test this PR.

**Comment by glebius** [comment]
@vstinner @nirs @zooba Guys, who is actually handling this request?

**Comment by vstinner** [comment]
> I'd really appreciate if this happens soon rather than wait for 14.0-RELEASE

Can't you have a downstream patch in FreeBSD to test your change before FreeBSD 14 is released?

That's what we do in Fedora to unblock issues before upstream accepts our patches (or when the change is merged but not released yet).

I would prefer to wait until FreeBSD 14 is released, before making changes to support it. By the way, Python 3.12 is not going to be released before October 2023 (in one year).

**Comment by vstinner** [comment]
> The constant wasn't defined before and it stays undefined. The old style socket could actually be used, just using 258 literal.

Can you please propose a PR just to add IPPROTO_DIVERT? I could test it immediately.

**Comment by glebius** [comment]
Status update: mmkay, I have added this patch to all python ports on FreeBSD. Vicotr, I hope you will merge it once FreeBSD 14.0-RELEASE is officially out.

**Review by vstinner — COMMENTED** [review]
Would it be possible to somehow document these constants? See for example:
https://docs.python.org/dev/library/socket.html#socket.AF_RDS

Document that they are new in Python 3.12.

**Comment by vstinner** [comment]
I understand that you have this downstream patch in FreeBSD CURRENT for a few weeks/months. I'm now fine with adding these FreeBSD 14 constants (once the review will be done).

**Comment by glebius** [comment]
@vstinner updated socket module documentation as you suggested. Let me know if I need anything else to be done to get this in.

**Comment by glebius** [comment]
@vstinner We are entering FreeBSD 14 release cycle. I'd like to get this in python before FreeBSD 14.0 is released. Let me know how I can help to achieve that.

**Review by vstinner — APPROVED** [review]
LGTM. Thanks for adding the documentation.

**Comment by vstinner** [comment]
Better late than never, I merged your PR. Thanks for your contribution.

**Comment by koobs** [comment]
> Better late than never, I merged your PR. Thanks for your contribution.

Thanks Victor! ❤️

**Pull request merged**
